crates/metaslang/cst/src/text_index.rs

```diff
@@ -1,22 +1,46 @@
 use std::fmt::Display;
-use std::ops::{Add, AddAssign, Range, Sub, SubAssign};
+use std::ops::{Add, AddAssign, Range};
 
 use serde::Serialize;
 
-#[derive(Default, Copy, Clone, PartialEq, Eq, Debug, Serialize)]
+#[derive(Default, Hash, Copy, Clone, PartialEq, Eq, Debug, Serialize)]
 pub struct TextIndex {
     pub utf8: usize,
     pub utf16: usize,
-    pub char: usize,
+    pub line: usize,
+    pub column: usize,
 }
 
 impl TextIndex {
-    /// Shorthand for `TextIndex { utf8: 0, utf16: 0, char: 0 }`.
+    /// Shorthand for `TextIndex { utf8: 0, utf16: 0, line: 0, char: 0 }`.
     pub const ZERO: TextIndex = TextIndex {
         utf8: 0,
         utf16: 0,
-        char: 0,
+        line: 0,
+        column: 0,
     };
+
+    #[inline]
+    pub fn advance(&mut self, c: char, next: Option<&char>) {
+        self.utf8 += c.len_utf8();
+        self.utf16 += c.len_utf16();
+        if c == '\n'
+            // line separator
+            || c == '\u{2028}'
+            // paragraph separator
+            || c == '\u{2029}'
```

Comment:
What the language constitutes as a newline affects parsing but in the case of calculating a correct line number we must make sure that this will be consistent across multiple text editors or other tools that might ingest that data. 

Both Vim and VS Code treat LS and PS as whitespace; we'd need to test more editors, but I think we should probably only account for LF/CR/CRLF here.

Reply:
In order to be consistent with other tools, that obey the official solidity definition of a line break, I'm choosing to use the solidity definition here. This is not consistent across languages, nor across editors or other text tools (`wc` ?), but I think here we should be merely pragmatic.

Reply:
ECMAScript also uses these and I was curious what the behaviour is, so I went ahead and tried starting the code with LS and curiously enough, the `tsc` really does treat it as a logical newline in the diagnostics:

![image](https://github.com/NomicFoundation/slang/assets/3093213/753446d9-b95a-46aa-9098-ab75858a620a)
![image](https://github.com/NomicFoundation/slang/assets/3093213/57a353f7-6e72-471c-9c83-8db6b2d89c54)

While it seems peculiar, I rest my case then but it'd probably be good to add a comment that it explicitly matches Solidity to clear any possible confusion; long-term, it should be dependent on the language definition but let's not go there now :sweat_smile:

Reply:
'rest my case' would mean that you were proved correct :), but I think you mean that you accept we should use `LS` and `PS` as line break?

I agree we should use the language def, but that will never be explicitly synced with the editor or tool behaviour (apart from in metaslang!), it will only coincidentally match until the world 100% conforms to unicode.

Reply:
Ah, my bad then! Meant to say that I've resolved my concern.

I went down the rabbit hole while I was looking at https://github.com/NomicFoundation/slang/pull/918#discussion_r1604914705 to see if how they treat bare `\r` and in general, I found that `solc` does some really funny stuff as usual:

I'll start with the fact that despite defining the PS/LS/NEL sequences as terminating the string literals wrt. lines, you can't reasonably use PS/LS in general in the code, because (at least when I tested it) it just throws a parser error anyway :sweat_smile: .

They use byte offsets almost exclusively, even as the range for the errors in the Standard JSON output, so they escape the problem of how to interpret lines in the majority of cases. However, when they need the lines as well, they use [`CharStream::translatePositionToLineColumn`](https://github.com/ethereum/solidity/blob/ae40e1a04acacd639fed0dfcb3dd1cdc11406044/liblangutil/CharStream.cpp#L102), which defines the lines as... literally just the count of the `\n` characters. That's what their built-in LSP uses, so they get correct lines in the editor, despite the presence of PS/LS (ignoring the fact that it means a parse error anyway).

So, with all that said, for anything that needs lines, I'd say that we probably should just mirror what solc does and only count `\n`, because it's simpler and what they do (and PS/LS is unusable). What do you think?

Reply:
using just one char and ignoring `\r\n` would have been simpler, but I've done the work now.

I'm wondering if this is a case where we do the *right* thing, rather than the *solc* thing?

Reply:
I'm worried that doing the "right" thing here is actually not a pragmatic choice, i.e. we will get incorrect line numbers in almost all of the editors. However, given that any Solidity code using LS/PS seems to be invalid anyway, I don't want to block this as hopefully this is a non-issue anyway.